File: com.se.simplicity/src/com/se/simplicity/rendering/Light.java
/*
    This file is part of The Simplicity Engine.

    The Simplicity Engine is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published
    by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

    The Simplicity Engine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with The Simplicity Engine. If not, see <http://www.gnu.org/licenses/>.
 */
package com.se.simplicity.rendering;

import com.se.simplicity.scenegraph.Node;
import com.se.simplicity.vector.TransformationMatrixf;

/**
 * <p>
 * A light within a {@link com.se.simplicity.scenegraph.SceneGraph SceneGraph}.
 * </p>
 * 
 * @author Gary Buyn
 */
public interface Light
{
    /**
     * <p>
     * Applies this <code>Light</code> to the rendering environment.
     * </p>
     */
    void apply();

    /**
     * <p>
     * Retrieves the ambient component of this <code>Light</code>.
     * </p>
     * 
     * @return The ambient component of this <code>Light</code>.
     */
    float[] getAmbientLight();

    /**
     * <p>
     * Retrieves the diffuse component of this <code>Light</code>.
     * </p>
     * 
     * @return The diffuse component of this <code>Light</code>.
     */
    float[] getDiffuseLight();

    /**
     * <p>
     * Retrieves the lighting mode used to render a {@link com.se.simplicity.scenegraph.SceneGraph SceneGraph}.
     * </p>
     * 
     * @return The lighting mode used to render a {@link com.se.simplicity.scenegraph.SceneGraph SceneGraph}.
     */
    LightingMode getLightingMode();

    /**
     * <p>
     * Retrieves the {@link com.se.simplicity.scenegraph.Node Node} that represents this <code>Light</code>'s location and orientation.
     * </p>
     * 
     * @return The {@link com.se.simplicity.scenegraph.Node Node} that represents this <code>Light</code>'s location and orientation.
     */
    Node getNode();

    /**
     * <p>
     * Retrieves the specular component of this <code>Light</code>.
     * </p>
     * 
     * @return The specular component of this <code>Light</code>.
     */
    float[] getSpecularLight();

    /**
     * <p>
     * Retrieves the inverted absolute transformation for the {@link com.se.simplicity.scenegraph.Node Node} of this <code>Light</code>.
     * </p>
     * 
     * @return The inverted absolute transformation for the {@link com.se.simplicity.scenegraph.Node Node} of this <code>Light</code>, or null if the
     * {@link com.se.simplicity.scenegraph.Node Node} does not exist.
     */
    TransformationMatrixf getTransformation();

    /**
     * <p>
     * Initialises this <code>Light</code>.
     * </p>
     */
    void init();

    /**
     * <p>
     * Determines the initialisation status of this <code>Light</code>.
     * </p>
     * 
     * @return True if this <code>Light</code> has been initialised, false otherwise.
     */
    boolean isInitialised();

    /**
     * <p>
     * Sets the ambient component of this <code>Light</code>.
     * </p>
     * 
     * @param ambientLight The ambient component of this <code>Light</code>.
     */
    void setAmbientLight(float[] ambientLight);

    /**
     * <p>
     * Sets the diffuse component of this <code>Light</code>.
     * </p>
     * 
     * @param diffuseLight The diffuse component of this <code>Light</code>.
     */
    void setDiffuseLight(float[] diffuseLight);

    /**
     * <p>
     * Sets the initialisation status of this <code>Light</code>.
     * </p>
     * 
     * @param isInitialised The initialisation status of this <code>Light</code>.
     */
    void setInitialised(boolean isInitialised);

    /**
     * <p>
     * Sets the lighting mode used to render a {@link com.se.simplicity.scenegraph.SceneGraph SceneGraph}.
     * </p>
     * 
     * @param lightingMode The lighting mode used to render a {@link com.se.simplicity.scenegraph.SceneGraph SceneGraph}.
     */
    void setLightingMode(LightingMode lightingMode);

    /**
     * <p>
     * Sets the {@link com.se.simplicity.scenegraph.Node Node} that represents this <code>Light</code>'s location and orientation.
     * </p>
     * 
     * @param node The {@link com.se.simplicity.scenegraph.Node Node} that represents this <code>Light</code>'s location and orientation.
     */
    void setNode(Node node);

    /**
     * <p>
     * Sets the specular component of this <code>Light</code>.
     * </p>
     * 
     * @param specularLight The specular component of this <code>Light</code>.
     */
    void setSpecularLight(float[] specularLight);
}

File: simplicity/src/main/c++/simplicity/rendering/Light.cpp
/*
* Copyright © 2014 Simple Entertainment Limited
*
* This file is part of The Simplicity Engine.
*
* The Simplicity Engine is free software: you can redistribute it and/or modify it under the terms of the GNU General
* Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option)
* any later version.
*
* The Simplicity Engine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
* implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public License along with The Simplicity Engine. If not, see
* <http://www.gnu.org/licenses/>.
*/
#include "Light.h"

using namespace std;

namespace simplicity
{
	Light::Light(const string& name) :
		active(false),
		ambient(),
		attenuation(),
		diffuse(),
		direction(),
		name(name),
		range(100.0f),
		specular(),
		strength(1.0f)
	{
		activate();
	}

	void Light::activate()
	{
		active = true;
	}

	void Light::apply(Pipeline& pipeline, const Vector3& position)
	{
		pipeline.set(name + "Light", "attenuation", attenuation);
		pipeline.set(name + "Light", "direction", direction);
		pipeline.set(name + "Light", "position", position);
		pipeline.set(name + "Light", "range", range);
		pipeline.set(name + "Light", "specular", specular);
		pipeline.set(name + "Light", "strength", strength);

		if (active)
		{
			pipeline.set(name + "Light", "ambient", ambient);
			pipeline.set(name + "Light", "diffuse", diffuse);
			pipeline.set(name + "Light", "specular", specular);
		}
		else
		{
			pipeline.set(name + "Light", "ambient", Vector4(0.0f, 0.0f, 0.0f, 1.0f));
			pipeline.set(name + "Light", "diffuse", Vector4(0.0f, 0.0f, 0.0f, 1.0f));
			pipeline.set(name + "Light", "specular", Vector4(0.0f, 0.0f, 0.0f, 1.0f));
		}
	}

	void Light::deactivate()
	{
		active = false;
	}

	const Vector4& Light::getAmbient() const
	{
		return ambient;
	}

	const Vector3& Light::getAttenuation() const
	{
		return attenuation;
	}

	const Vector4& Light::getDiffuse() const
	{
		return diffuse;
	}

	const Vector3& Light::getDirection() const
	{
		return direction;
	}

	const string& Light::getName() const
	{
		return name;
	}

	float Light::getRange() const
	{
		return range;
	}

	const Vector4& Light::getSpecular() const
	{
		return specular;
	}

	float Light::getStrength() const
	{
		return strength;
	}

	bool Light::isActive() const
	{
		return active;
	}

	void Light::setAmbient(const Vector4& ambient)
	{
		this->ambient = ambient;
	}

	void Light::setAttenuation(const Vector3& attenuation)
	{
		this->attenuation = attenuation;
	}

	void Light::setDiffuse(const Vector4& diffuse)
	{
		this->diffuse = diffuse;
	}

	void Light::setDirection(const Vector3& direction)
	{
		this->direction = direction;
	}

	void Light::setRange(float range)
	{
		this->range = range;
	}

	void Light::setSpecular(const Vector4& specular)
	{
		this->specular = specular;
	}

	void Light::setStrength(float strength)
	{
		this->strength = strength;
	}
}
